src/main/java/com/sps/Servlets/NewRecipeServlet.java

```diff
@@ -0,0 +1,138 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.gson.Gson;
+import java.util.Collection;
+import java.util.List;
+import java.util.HashSet;
+import java.util.ArrayList;
+import com.google.sps.data.Recipe;
+import com.google.sps.data.Step;
+
+@WebServlet("/new-recipe")
+public class NewRecipeServlet extends HttpServlet {
+
+  private DatastoreService datastore;
+  private final String TAG = "tag";
+  private final String INGREDIENT = "ingredient";
+  private final String STEP = "step";
+
+  @Override
+  public void init() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /*
+   * When a spin-off is created, this GET method gets the original recipe's data.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String keyString = request.getParameter("key");
+    System.out.println(keyString);
+    Entity recipeEntity = null;
+    try {
+      recipeEntity = datastore.get(KeyFactory.stringToKey(keyString));
+    } catch (EntityNotFoundException e) {
+      e.printStackTrace();
+      return;
+    }
+    Recipe original = entityToRecipe(recipeEntity);
+    response.setContentType("application/json;");
+    response.getWriter().println(convertToJsonUsingGson(original));  
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Collection<String> searchStrings = new HashSet<>();
+    String name = request.getParameter("name");
+    searchStrings.add(name.toUpperCase());
+    String description = request.getParameter("description");
+    EmbeddedEntity tags = getParameters(request, TAG, searchStrings);
+    EmbeddedEntity ingredients = getParameters(request, INGREDIENT, searchStrings);
+    EmbeddedEntity steps = getParameters(request, STEP, null);
+
+    Entity recipe = new Entity("Recipe");
+    recipe.setProperty("name", name);
+    recipe.setProperty("description", description);
+    recipe.setProperty("tags", tags);
+    recipe.setProperty("ingredients", ingredients);
+    recipe.setProperty("steps", steps);
+    recipe.setProperty("search-strings", new ArrayList<String>(searchStrings));
+    datastore.put(recipe);
+
+    response.sendRedirect("/edit-recipe.html");
+  }
+
+  /**
+   * Gets the parameters for fields that have different numbers of parameters from recipe to recipe.
+   * For example, one recipe may have 2 ingredients, while another may have 20.
+   * This method ensures that all tags, ingredients, and steps are recorded, no matter how many of each a recipe has.
+   */
+  private EmbeddedEntity getParameters(HttpServletRequest request, String type, Collection<String> searchStrings) {
```

Comment:
Clarify what type is in the javadoc; I don't think its immediately clear enough to not be added in the javadoc.

Reply:
I end up resolving these in a later PR!

Reply:
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
ok